File: app/src/main/java/au/com/codeka/weather/model/HourlyForecast.java
package au.com.codeka.weather.model;

/**
 * Like {@link Forecast}, but for an hourly forecast, slightly simplified.
 */
public class HourlyForecast {
  private int hour;
  private String shortDescription;
  private double temperature;
  private WeatherIcon icon;
  private double qpfMillimeters;

  /**
   * Gets the "hour" of this forecast, in 24 hour time. We expect it to be the current day or
   * tomorrow if it passes midnight.
   */
  public int getHour() {
    return hour;
  }

  public String getShortDescription() {
    return shortDescription;
  }

  public double getTemperature() {
    return temperature;
  }

  public double getQpfMillimeters() {
    return qpfMillimeters;
  }

  public WeatherIcon getIcon() {
    return icon;
  }

  public static class Builder {
    private HourlyForecast forecast;

    public Builder() {
      forecast = new HourlyForecast();
    }

    public Builder setHour(int hour) {
      forecast.hour = hour;
      return this;
    }

    public Builder setShortDescription(String shortDescription) {
      forecast.shortDescription = shortDescription;
      return this;
    }

    public Builder setQpfMillimeters(double qpf) {
      forecast.qpfMillimeters = qpf;
      return this;
    }

    public Builder setTemperature(double temp) {
      forecast.temperature = temp;
      return this;
    }

    public Builder setIcon(WeatherIcon icon) {
      forecast.icon = icon;
      return this;
    }

    public HourlyForecast build() {
      return forecast;
    }
  }
}

File: app/src/main/java/au/com/codeka/weather/model/HourlyForecast.kt
package au.com.codeka.weather.model

/**
 * Like [Forecast], but for an hourly forecast, slightly simplified.
 */
class HourlyForecast {
  /**
   * Gets the "hour" of this forecast, in 24 hour time. We expect it to be the current day or
   * tomorrow if it passes midnight.
   */
  var hour = 0
    private set
  var shortDescription: String? = null
    private set
  var temperature = 0.0
    private set
  var icon: WeatherIcon? = null
    private set
  var qpfMillimeters = 0.0
    private set

  class Builder {
    private val forecast: HourlyForecast
    fun setHour(hour: Int): Builder {
      forecast.hour = hour
      return this
    }

    fun setShortDescription(shortDescription: String?): Builder {
      forecast.shortDescription = shortDescription
      return this
    }

    fun setQpfMillimeters(qpf: Double): Builder {
      forecast.qpfMillimeters = qpf
      return this
    }

    fun setTemperature(temp: Double): Builder {
      forecast.temperature = temp
      return this
    }

    fun setIcon(icon: WeatherIcon?): Builder {
      forecast.icon = icon
      return this
    }

    fun build(): HourlyForecast {
      return forecast
    }

    init {
      forecast = HourlyForecast()
    }
  }
}